### PR TITLE
HSEARCH-2477 Shard filtering doesn't work with Elasticsearch queries

### PR DESCRIPTION
**Pull request by yrodiere**
https://hibernate.atlassian.net/browse/HSEARCH-2477

~~Not to be merged immediately, I am going to do a release.~~ => Done

### CODE REVIEW

File: engine/src/main/java/org/hibernate/search/query/engine/impl/LuceneHSQuery.java [hunk 1]
```diff
@@ -282,6 +279,25 @@ protected void extractFacetResults() {
...
 
+	@Override
+	protected List<IndexManager> getIndexManagers(EntityIndexBinding binding) {
+		List<IndexManager> indexManagers = super.getIndexManagers( binding );
+
+		for ( IndexManager indexManager : indexManagers ) {
+			if ( !( indexManager instanceof DirectoryBasedIndexManager ) ) {
+				//TODO clarify intent:
+				// A) improve the error message so that people understand what they should do
+				// B) Is the point really to not support "directory-based" or rather non-Lucene native based ones?
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/query/engine/impl/LuceneHSQuery.java]: Hmmm, it looks weird to limit this to directory based indexes. I would expect this method to work even for RAM based indexes.

I then realized that it was already existing code but I thought I would mention it anyway.

File: engine/src/main/java/org/hibernate/search/query/engine/impl/AbstractHSQuery.java [hunk 2]
```diff
@@ -530,6 +535,25 @@ private FacetMetadata findFacetMetadata(FacetingRequest facetingRequest, Iterabl
 		return facetMetadata;
 	}
 
+	protected List<IndexManager> getIndexManagers(EntityIndexBinding binding) {
+		FullTextFilterImplementor[] fullTextFilters = getFullTextFilterImplementorsAsArray();
+		List<IndexManager> indexManagers = Arrays.asList( binding.getSelectionStrategy().getIndexManagersForQuery( fullTextFilters ) );
+		return indexManagers;
+	}
+
+	private FullTextFilterImplementor[] getFullTextFilterImplementorsAsArray() {
```
Comment (gsmet) [inline engine/src/main/java/org/hibernate/search/query/engine/impl/AbstractHSQuery.java]: Is the fact that it returns an Array really important? I wouldn't include it in the method name.

And I must admit I preferred the old method name altogether (`getFullTextFilter()`). It's more to the point and doesn't add useless information.

File: engine/src/main/java/org/hibernate/search/query/engine/impl/LuceneHSQuery.java (hunk 1 above)
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/query/engine/impl/LuceneHSQuery.java]: RAM-based indexes are directory-based: they use a `RAMDirectoryProvider`. See `BaseDirectoryProviderService`.

Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/query/engine/impl/LuceneHSQuery.java]: But you're right: this comment should disappear eventually (why the hell create a TODO in the code without a corresponding ticket???). I created [HSEARCH-2724](https://hibernate.atlassian.net/browse/HSEARCH-2724).

File: engine/src/main/java/org/hibernate/search/query/engine/impl/AbstractHSQuery.java (hunk 2 above)
Comment (yrodiere) [inline engine/src/main/java/org/hibernate/search/query/engine/impl/AbstractHSQuery.java]: FullTextFilter is much too vague, we could be returning half a dozen different things, and none of them would be handled the same way:

 * `Query`
 * `Filter`
 * `FullTextFilterImplementor`
 * `FullTextFilterDef` (!)
 * `FullTextFilter` (!)
 * etc.

I'll remove the "asArray" suffix, though.

**Comment by yrodiere** [comment]
@gsmet Thanks, I removed the "asArray" suffix. I will merge and backport as soon as the CI build succeeds.

**Comment by yrodiere** [comment]
Merged and backported to 5.6 and 5.7.

**Pull request merged**